cpp/common/src/codingstandards/cpp/alertreporting/DeduplicateMacroResults.qll

```diff
@@ -0,0 +1,379 @@
+ * ```
+ *
+ * This module exposes the following classes:
+ *  - `PrimaryMacroSameResultElementInAllInvocations extends Macro`: Every invocation of this macro
+ *    generates an `InvalidFoo` which stringifies to the same thing. Use the method
+ *    `getResultElementDescription()` to get that shared string.
+ *  - `PrimaryMacroDifferentResultElementInAllInvocations extends Macro`: Every invocation of this
+ *    macro generates an `InvalidFoo`, but they do not all stringify to the same thing. Use the
+ *    method `getExampleResultElement()` to get an *single* example `InvalidFoo` to help users fix
+ *    and understand the issue.
+ *  - `IsolatedMacroExpansionWithResultElement extends MacroInvocation`: An invocation of a macro
+ *    which in this particular instance generates an `InvalidFoo`, while other invocations of the
+ *    same macro do not.
+ *
+ * The three above classes all attempt to resolve to the most *specific* macro to the issue at
+ * hand. For instance, if macro `M1` calls macro `M2` which expands to an `InvalidFoo`, then the
+ * problem may be with `M2` (it is the most specific macro call here), or the problem may be with
+ * `M2` (if all expansions of `M2` generate an `InvalidFoo` but not all expansions of `M1` do so).
+ *
+ * ## Generating Report Objects
+ *
+ * This module also can be used to more easily report issues across these cases, by implementing
+ * `MacroReportConfigSig` and importing `DeduplicateMacroResults::Report::ReportResultInMacro`.
+ *
+ * ```
+ *   module InvalidFooInMacroReportConfig implements MacroReportConfigSig<InvalidFoo> {
+ *
+ *     // ***Take care that usage of $@ is correct in the following predicates***!!!!
+ *     bindingset[description]
+ *     string getMessageSameResultInAllExpansions(Macro m, string description) {
+ *       result = "Macro " + m.getName() + " always has invalid foo " + description
+ *     }
+ *
+ *     string getMessageVariedResultInAllExpansions(Macro m) {
+ *       result = "Macro " + m.getName() + " always has invalid foo, for example '$@'."
+ *     }
+ *
+ *     string getMessageResultInIsolatedExpansion(InvalidFoo foo) {
+ *       result = "Invocation of macro $@ has invalid foo '" + foo.getName() + "'."
+ *     }
+ *   }
+ *
+ *   import DeduplicateFooInMacros::Report<InvalidFooInMacroReportConfig> as Report;
+ *
+ *   from Report::ReportResultInMacro report
+ *     where not excluded(report.getPrimaryElement(), ...)
+ *   select report.getPrimaryElement(), report.getMessage(), report.getOptionalPlaceholderLocation(),
+ *     report.getOptionalPlaceholderMessage()
+ * ```
+ *
+ * Note that this does *not* (currently) generate a result for elements not contained by a macro.
+ * To do report such cases, either add support for that in this module, or write a separate query
+ * that reports `InvalidFoo` cases where not `.isInMacroExpansion()`.
+ */
+module DeduplicateMacroResults<
```

Comment:
I believe we could use this module in `A5-2-2`, which already has some manual handling for duplicate macro results, but could benefit from the improvements in this module. I don't think we need to do it in this PR, but should open an issue to track this as a follow up item.

Reply:
Filed https://github.com/github/codeql-coding-standards/issues/839